Enhanced unified video streaming app + integration testing

cdinea opened the pull request:
[to be added]

tbirdso reviewed (commented):
Initial review pass focused on integration test design, please see feedback

cdinea commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,156 @@
+cmake_minimum_required(VERSION 3.20)
+project(video_streaming_demo_enhanced LANGUAGES NONE)
+
+# Check if this application should be built
+if(NOT APP_video_streaming_demo_enhanced)
+  return()
+endif()
```
thank you for the feedback @tbirdso - I believe I ran into build errors when trying to use add_holohub_application , would you consider this to be a P0 item and needs revisiting?

tbirdso replied:
Yes, please reach out internally for help if you encounter issues with HoloHub infrastructure.

cdinea commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,156 @@
+                 echo '✓ Client is still running - connection successful'
+                 CLIENT_SUCCESS=1
+
+                 # Stop client gracefully
+                 echo 'Stopping client...'
+                 kill -TERM \$CLIENT_PID 2>/dev/null || true
```
thank you @tbirdso - not sure how else we could stop a client process to stop running in a test environment.

cdinea commented on applications/video_streaming_demo_enhanced/Dockerfile:
```diff
@@ -75,9 +97,8 @@ WORKDIR /workspace/holohub
-
-# Enable autocomplete
-RUN echo "source /etc/bash_completion" >> ~/.bashrc
+ENV PYTHONPATH=/workspace/holohub/applications/streaming_server_demo/python:${PYTHONPATH}
```
thank you @tbirdso - PYTHONPATH deleted, already set in a previous line

cdinea commented on applications/video_streaming_demo_enhanced/README.md:
````diff
@@ -0,0 +1,670 @@
+
+### Continuous Integration
+
+The integration test is designed for CI/CD pipelines:
+
+```bash
+# CI-friendly command with timeout and exit codes
+timeout 300 ./applications/video_streaming_demo_enhanced/integration_test.sh
+echo "Integration test exit code: $?"
+```
````
hi @tbirdso - based on the integration test we outlined, my understanding is that only the apps will be run with ./holohub test

`# launch server
./holohub test --docker-opts='-e EnableHybridMode=1' streaming_server_demo_04_80_tensor_streaming --language cpp 2>&1 > streamignserver.log &  # here I am using '>' instead of '| tee' to not showing anything in the terminal

sleep 10

./holohub test --docker-opts='-e EnableHybridMode=1 --device /dev/video0' streaming_client_demo_04_80_streaming --language cpp --config streaming_client_demo.yaml 2>&1 > streamingclient.log &  # here I am using '>' instead of '| tee' to not showing anything in the terminal

sleep 30

# wait for both server and client to terminates (timeout = 10 secs)
wait -t 10

kill -9 %1 %2

# check the log files
cat streamingserver.log
cat streamingclient.log`


I need to scope out what it takes for ./holohub test to be able to run nested tests with ./holohub test

tbirdso commented on applications/video_streaming_demo_enhanced/README.md:
```diff
@@ -0,0 +1,670 @@
+
+The unified application provides both client and server as separate components:
+
+**⚠️ Important: Enhanced applications require Holoscan SDK 3.5.0. Use the base image parameter to ensure compatibility.**
```
Got it, thanks for the explanation. Agreed that documenting and pinning to 3.5.0 is sufficient.

tbirdso commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,236 @@
+             fi
+
+             echo ''
+             echo '=== Test Results Summary ==='
+             echo \"Server checks passed: \$SERVER_SUCCESS/4 (minimum required: 3)\"
+             echo \"Client checks passed: \$CLIENT_SUCCESS/3 (minimum required: 2)\"
+
+             # Determine overall success
+             # Require at least 3/4 server checks and 2/3 client checks
+             if [ \$SERVER_SUCCESS -ge 3 ] && [ \$CLIENT_SUCCESS -ge 2 ]; then
+               STREAMING_SUCCESS=1
```
What is the justification for these minimum bounds on acceptance?

It seems like we should always expect 100% pass, anything less is a test failure.

Also, I count 5 server checks.. the implementation with hardcoded constants here will not scale if new checks are added and not accounted for on this line. It seems like the current implementation claims 80% minimum bound for the server but is actually 60%.

Please update the test logic so that the existence of _any_ failure will result in an overall test failure.

tbirdso commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,236 @@
+             fi
+             
+             # Check for frame processing
+             if grep -q 'Processing UNIQUE frame' \$SERVER_LOG; then
+               FRAME_COUNT=\$(grep -c 'Processing UNIQUE frame' \$SERVER_LOG)
+               echo \"✓ Server: Processed \$FRAME_COUNT unique frames\"
```
What is the expected lower and upper bound on the frame count? Current implementation could pass with 0 unique frames, which seems not expected

tbirdso commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,236 @@
+               echo '✗ Server: No frames processed'
+             fi
+             
+             # Check for frame stats (optional but good indicator)
+             if grep -q 'Frame Processing Stats' \$SERVER_LOG; then
+               echo '✓ Server: Frame processing statistics logged'
+               SERVER_SUCCESS=\$((SERVER_SUCCESS + 1))
+             fi
```
The integration test should not have optional checks, we should have clear expectations and acceptance criteria for deterministic test behavior. Please clarify whether or not this is expected output for the integration test, and if the output is expected then please make the check required for the test to pass.

tbirdso commented on operators/video_streaming/streaming_client_enhanced/streaming_client.cpp:
```diff
@@ -0,0 +1,1601 @@
+    shape_str += "]";
+    HOLOSCAN_LOG_INFO("Tensor shape={}", shape_str);
+
+    // 🔧 ENHANCED DEBUG: Write tensor to disk before processing - standardized to every 10th frame
+    debug_frame_counter++;
+    if (debug_frame_counter % 10 == 0) {  // Save every 10th frame to match server frequency
+      HOLOSCAN_LOG_INFO("💾 DEBUG: Writing input tensor to disk (frame {})", debug_frame_counter);
+      writeTensorToDisk(tensor, "client_input_tensor", debug_frame_counter);
+    }
```
Please guard debug functionality behind a disabled-by-default flag, or remove entirely

tbirdso commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,236 @@
+             echo ''
+             echo '=== Verifying Client Logs ==='
+             
+             # Check for successful frame sending
```
Thank you for adding regex checks. It looks like the test is still missing regex coverage for client frame reception, would you please add that as well?

tbirdso commented on applications/video_streaming_demo_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,236 @@
+               echo '✗ Server: Downstream connection not established'
+             fi
+             
+             # Check for frame processing
+             if grep -q 'Processing UNIQUE frame' \$SERVER_LOG; then
```
This check looks good and verifies that the `StreamingServerUpstreamOp::compute` method has run.

Please also add a regex check to verify that frames are handled in `StreamingServerDownstreamOp`.

tbirdso reviewed (commented):
Thank you for adding regex checks in the integration test, added feedback.

bhashemian commented:
cursor review

bhashemian commented on operators/video_streaming/streaming_client_enhanced/streaming_client.cpp:
```diff
@@ -0,0 +1,1601 @@
+
+          send_success = client_->sendFrame(frame);
+          if (send_success) {
+            HOLOSCAN_LOG_INFO("✅ Frame sent successfully on attempt {}",
```
I am not so sure of using icons and emojis in the code. Having a series of tests that provides a summary of what has passed or failed would be much cleaner and easier to maintain? @tbirdso what do you think?

bhashemian reviewed (changes requested):
@cdinea thanks so much for this contribution. I have left few comments.

One challenge with this PR is the excessive logging information and unnecessary comments that reduce code readability. I would greatly appreciate it if you could make them concise and remove any unnecessary logging and comments.

cdinea commented on operators/video_streaming/streaming_client_enhanced/CMakeLists.txt:
```diff
@@ -0,0 +1,282 @@
+  ${CMAKE_BINARY_DIR}/operators/streaming_client/libcrypto.so.3
+  COMMAND ${CMAKE_COMMAND} -E copy_if_different
+  ${STREAMINGCLIENT_DIR}/lib/${ARCH_DIR}/libcudart.so.12.0.107
+  ${CMAKE_BINARY_DIR}/operators/streaming_client/libcudart.so.12.0.107
```
thank you for the feedback @bhashemian - these libraries came bundled together with the streaming and any modification would risk ABI compatibility issues and potential runtime failures.

cdinea commented on applications/video_streaming_demo_enhanced/video_streaming_demo_client/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,109 @@
+             PATHS "/opt/nvidia/holoscan" "/workspace/holoscan-sdk/install")
+
+# Find the streaming client operator directory
+set(OPERATOR_CLIENT_DIR "${CMAKE_SOURCE_DIR}/operators/video_streaming/streaming_client_enhanced/holoscan_client_cloud_streaming/lib/x86_64")
```
Good catch! This is indeed an oversight. The libraries are actually provided for both x86_64 and aarch64 architectures in the NGC package. The server CMakeLists.txt already has proper architecture detection, but I missed implementing it in the client CMakeLists.txt. I'll update it to match the server's approach to support both architectures properly.

bhashemian replied:
@cdinea any update on this?

bhashemian commented:
Thanks @cdinea for addressing the comments. @tbirdso, do you have any other comments here? If not, I’ll locally test and merge this PR.

bhashemian commented:
@cdinea running the server enhanced applications failed with the following error:
```
FileNotFoundError: [Errno 2] No such file or directory: '/workspace/holohub/build/video_streaming_demo_server/applications/video_streaming_demo_enhanced/video_streaming_demo_server/cpp/streaming_server_demo_enhanced_wrapper.sh'
```

cdinea commented:
thank you for the feedback @bhashemian - I have tested the app yesterday but I will test again and report back - it looks like one of the files might have got lost during rebase or cherry picking.

cdinea commented:
@bhashemian  I updated the READMEs- all the required files are present, but by changing the build structure and use add_holohub_applicationinstead of add_subdirectory for the parent app, the holohub run commands have changed and some READMEs have not been updated with the new holohub run commands.I have tested on my end , please let me know if you encounter any issues.I   also added a screenshot in the main app README

cdinea commented:
FYI Once this PR is merged - I will rebase [this](https://github.com/nvidia-holoscan/holohub/pull/1152) PR that has the Python bindings and one integration test and open it for review

cdinea commented on applications/video_streaming_demo_enhanced/screenshot_streaming_demo.png:
I pushed the commit with the screenshot for the Holoviz window

cdinea commented on operators/video_streaming/metadata.json:
```diff
@@ -0,0 +1,51 @@
+        },
+        "language": ["C++"],
+        "platforms": ["x86_64", "aarch64"],
+        "tags": ["Video", "Streaming", "Client", "Server", "Networking", "Cloud", "Bidirectional"],
```
@bhashemian is this feedback valid and needs to be addressed?

bhashemian replied:
Yes but I'll take care of that.

cdinea replied:
is this feedback related to [this ](https://github.com/nvidia-holoscan/holohub/pull/1165)PR which has been merged ?

it looks like there might be a bug, since "Video" is part of the approved list 
"Video" ✅ (approved)

But it might be checking: "Networking" instead of "Networking and Distributed Computing" ❌
Or checking if "Cloud" is a valid category ?

probably this PR introduces a list that does not match the current tags and an option would be to remove non-category tags like "Client", "Server", "Cloud", "Bidirectional" - maybe it is good to document this somewhere so our developers can chose only from the approved list

cdinea commented:
@bhashemian I am rebasing the branch now and will ping once done

bhashemian commented on applications/video_streaming_demo_enhanced/video_streaming_demo_client/cpp/CMakeLists.txt:
```diff
@@ -0,0 +1,118 @@
+#   set_tests_properties(streaming_client_demo_enhanced_cpp_test PROPERTIES
+#                        PASS_REGULAR_EXPRESSION "StreamingClient created successfully"
+#                        FAIL_REGULAR_EXPRESSION "[^a-z]Error;ERROR;Failed")
+# endif()
```
@cdinea Can we remove this commented block?
